### ISSUE TITLE
ORM Collection OneToMany cache issue on update

### ISSUE THREAD

**Issue by mpoiriert** [doctrine/doctrine2#5821]
Following

http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/second-level-cache.html

Under 33.6.1. Cache usage

There is a example of OneToMany collection cache (State->cities).

From this documentation

http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/unitofwork-associations.html

In a OneToMany the owner side is the inverse side (City in the example mention above).

From the UnitOfWork

``` PHP
 private function computeAssociationChanges($assoc, $value)
    {
        if ($value instanceof Proxy && ! $value->__isInitialized__) {
            return;
        }

        if ($value instanceof PersistentCollection && $value->isDirty()) {
            $coid = spl_object_hash($value);

            if ($assoc['isOwningSide']) {
                $this->collectionUpdates[$coid] = $value;
            }

            $this->visitedCollections[$coid] = $value;
        }
 ///...///
}
```

This mean the collection will not be added to the **collectionUpdates** since it's not the owning side.

Not being in the **collectionUpdates** the **AbstractCollectionPersister->update** will not get called so the cache entry will not be updated (or deleted) resulting of corrupted data in the cache.

It's all a big theory but looking in the code it make sense that this is my problem (unless there is something I didn't configure properly).

Clearing the cache after the flush does make all the cache entry getting updated and it's working properly...

Also I was trying to find automation test of those classes and I didn't find any, could I be mistaken or there is really no automated integration test for this ?

**Comment by janlanger** [comment]
I just came across same issue. Inverse side configured with cascade persist and orphan removal, but adding or removing of item from collection does not refresh collection cache.

Workaround is 
```php
$this->entityManager->getCache()->evictCollection(get_class($entity), 'property', $identifier);
```

**Comment by mpoiriert** [comment]
I am a bit late but were did you put this line ? Right now I have a listener of the OnFlush event, I was clearing everything a bit to evelly (the region namespace). Using your code will be more precise just want to make sure the "listener" approach is good.

It's really hacky but here is what I have right now:

```
/**
 * Auto clear cache system until a doctrine bug is fixed
 *
 * @see https://github.com/doctrine/doctrine2/issues/5821
 */
class ClearCacheListener
{
    public function onFlush(OnFlushEventArgs $args)
    {
        $unitOfWork = $args->getEntityManager()->getUnitOfWork();

        $reflectionObject = new \ReflectionObject($unitOfWork);
        $reflectionProperty = $reflectionObject->getProperty('visitedCollections');
        $reflectionProperty->setAccessible(true);
        $visitedCollections = $reflectionProperty->getValue($unitOfWork);

        $collections = array_merge(
            $unitOfWork->getScheduledCollectionDeletions(),
            $unitOfWork->getScheduledCollectionUpdates()
        );

        $evicted = [];
        foreach ($visitedCollections as $key => $visitedCollection) {
            if (array_key_exists($key, $collections)) {
                continue;
            }
            /* @var $visitedCollection \Doctrine\ORM\PersistentCollection */
            if (!$visitedCollection->isDirty()) {
                continue;
            }

            $entityPersister = $unitOfWork->getCollectionPersister($visitedCollection->getMapping());

            if (!$entityPersister instanceof CachedPersister) {
                continue;
            }

            if (in_array($entityPersister->getCacheRegion()->getName(), $evicted)) {
                continue;
            }

            $entityPersister->getCacheRegion()->evictAll();

            $evicted[] = $entityPersister->getCacheRegion()->getName();
        }
    }
}
```

**Comment by mpoiriert** [comment]
Using your method this is what I have right now

```
use Doctrine\ORM\Cache\Persister\CachedPersister;
use Doctrine\ORM\Event\OnFlushEventArgs;

/**
 * Auto clear cache system until a doctrine bug is fixed
 *
 * @see https://github.com/doctrine/doctrine2/issues/5821
 */
class ClearCacheListener
{
    public function onFlush(OnFlushEventArgs $args)
    {
        $entityManager = $args->getEntityManager();
        $unitOfWork = $args->getEntityManager()->getUnitOfWork();

        $reflectionObject = new \ReflectionObject($unitOfWork);
        $reflectionProperty = $reflectionObject->getProperty('visitedCollections');
        $reflectionProperty->setAccessible(true);
        $visitedCollections = $reflectionProperty->getValue($unitOfWork);

        $collections = array_merge(
            $unitOfWork->getScheduledCollectionDeletions(),
            $unitOfWork->getScheduledCollectionUpdates()
        );

        foreach ($visitedCollections as $key => $visitedCollection) {
            if (array_key_exists($key, $collections)) {
                continue;
            }
            /* @var $visitedCollection \Doctrine\ORM\PersistentCollection */
            if (!$visitedCollection->isDirty()) {
                continue;
            }
            $entityPersister = $unitOfWork->getCollectionPersister($visitedCollection->getMapping());

            if (!$entityPersister instanceof CachedPersister) {
                continue;
            }

            $entity = $visitedCollection->getOwner();
            $mapping = $visitedCollection->getMapping();
            $identifier = $unitOfWork->getEntityIdentifier($entity);
            $entityManager->getCache()->evictCollection(get_class($entity), $mapping['fieldName'], $identifier);
        }
    }
}
```

**Comment by janlanger** [comment]
@mpoiriert Well, we have my workaround directly where insert/update/delete is done. Not very nice, but its working.

I though about listener workaround, but didn't get to it, thanks for your code! But it's still bug and should be fixed in doctrine...

**Comment by mpoiriert** [comment]
yeah it should ! Just to let you know I have use the last version of what I paste above, it pass our integration test and we are releasing it in production this friday. My previous code was running since I open this issue but it was clearing the full region, ok for CMS data but less good for User Data...

**Comment by nicolas-cajelli** [comment]
I ran into the same issue today, and i was about to get my hands into the code when i found this issue (i googled a lot to get here).

It would be nice to get the fix, but in the meanwhile (since probably this will need lots of discussions), the documentation could be updated to reflect this (just for the sake of future innocents)

**Comment by lcobucci** [comment]
@mpoiriert could you check if this still happens on `2.6.x-dev`?

**Comment by janlanger** [comment]
@lcobucci this issue was fixed in 2.5.6 by https://github.com/doctrine/doctrine2/pull/6159

**Comment by lcobucci** [comment]
@janlanger thanks! Closing then.
